### ISSUE TITLE
Scientific notation

### ISSUE THREAD

**Issue by qwert2003** [uklimaschewski/EvalEx#51]
Scientific notation (```"2.2E-16"```) within expressions currently is not supported.

Example (calling EvalEx from BeanShell) ...
```
addClassPath("drawj2d.jar"); // BeanShell (bsh2) syntax
import com.udojava.evalex.*;
import java.math.BigDecimal;

String expr2 = "2.2E-16 * 10.2";
Expression expression = new Expression(expr2);
BigDecimal result = expression.eval();
```
... fails throwing an exception:
```
java.lang.RuntimeException: Unknown operator or function: E
        at com.udojava.evalex.Expression.shuntingYard(Expression.java:1149)
        at com.udojava.evalex.Expression.getRPN(Expression.java:1353)
        at com.udojava.evalex.Expression.eval(Expression.java:1166)
```

I once asked in the comments section of Udo's blog. The answer was:
> Scientific notation, this could get tricky. But I guess it is possible to implement in the tokenizer.

The java method call ```new BigDecimal("2.2E-16")``` does parse numbers in scientific notation. So it seems that it would be enough to tell the EvalEx tokenizer to understand the number in scientific notation as a single statement.
Trying ...
```
Index: com/udojava/evalex/Expression.java
===================================================================
--- com/udojava/evalex/Expression.java  (revision 263)
+++ com/udojava/evalex/Expression.java  (working copy)
@@ -593,7 +593,8 @@
                                ch = input.charAt(++pos);
                        }
                        if (Character.isDigit(ch)) {
-                               while ((Character.isDigit(ch) || ch == decimalSeparator)
+                               while ((Character.isDigit(ch) || ch == decimalSeparator
+                                                || ch == 'e' || ch == 'E')
                                                && (pos < input.length())) {
                                        token.append(input.charAt(pos++));
                                        ch = pos == input.length() ? 0 : input.charAt(pos);
@@ -1056,7 +1057,8 @@
                        return false;
                for (char ch : st.toCharArray()) {
                        if (!Character.isDigit(ch) && ch != minusSign
-                                       && ch != decimalSeparator)
+                                       && ch != decimalSeparator
+                                        && ch != 'e' && ch != 'E')
                                return false;
                }
                return true;
@@ -1398,6 +1400,7 @@
                        }
                        counter++;
                }
+                System.err.println("DEBUG: " + rpn.toString());
                if (counter > 1) {
                        throw new ExpressionException("Too many numbers or variables");
                } else if (counter < 1) {
```
... works for expressions like "2.2E16 * 10.2" but still fails for "2.2E-16".

Any idea how EvalEx can be told to recognize "2.2E-16" as a single number?

**Comment by squeek502** [comment]
Maybe store a boolean `lastCharWasE` and check for `ch == '-' && lastCharWasE`?

**Comment by qwert2003** [comment]
At the first glance the following diff seems to work.

```
Index: com/udojava/evalex/Expression.java
===================================================================
--- com/udojava/evalex/Expression.java	(revision 263)
+++ com/udojava/evalex/Expression.java	(working copy)
@@ -593,8 +593,11 @@
 				ch = input.charAt(++pos);
 			}
 			if (Character.isDigit(ch)) {
-				while ((Character.isDigit(ch) || ch == decimalSeparator)
-						&& (pos < input.length())) {
+				while ((Character.isDigit(ch) || ch == decimalSeparator
+                                                || ch == 'e' || ch == 'E'
+                                                || (ch == minusSign && token.length() > 0 
+                                                    && ('e'==token.charAt(token.length()-1) || 'E'==token.charAt(token.length()-1)))
+                                                ) && (pos < input.length())) {
 					token.append(input.charAt(pos++));
 					ch = pos == input.length() ? 0 : input.charAt(pos);
 				}
@@ -1056,7 +1059,8 @@
 			return false;
 		for (char ch : st.toCharArray()) {
 			if (!Character.isDigit(ch) && ch != minusSign
-					&& ch != decimalSeparator)
+					&& ch != decimalSeparator
+                                        && ch != 'e' && ch != 'E')
 				return false;
 		}
 		return true;
```

Could anybody test it? Does it break anything?

**Comment by qwert2003** [comment]
Pull request #52 enhanced.

**Comment by uklimaschewski** [comment]
Implemented with pr #52
